Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs

```diff
@@ -0,0 +1,259 @@
+        {
+            _amountCounterSpreading = amount;
+            _duration = duration;
+            _spreadDelay = spreadDelay;
+            _removeDelay = removeDelay;
+
+            // So the first square reacts immediately after spawning
+            _reactionTimer = ReactionDelay;
+            /*
+            The group takes amount*spreadDelay seconds to fully spread, same with fully disappearing.
+            The outer squares will last duration seconds.
+            The first square will last duration + how many seconds the group takes to fully spread and fully disappear, so
+            it will last duration + amount*(spreadDelay+removeDelay).
+            Thus, the average lifetime of the smokes will be (outerSmokeLifetime + firstSmokeLifetime)/2 = duration + amount*(spreadDelay+removeDelay)/2
+            */
+            _averageExposures = (duration + amount * (spreadDelay+removeDelay) / 2)/ReactionDelay;
+        }
+
+        public void InceptionUpdate(float frameTime)
+        {
+            _group.RemoveWhere(effect => effect.Deleted);
+            if (_group.Count == 0)
+                return;
+
+            // Make every outer square from the group spread
+            if (_amountCounterSpreading != 0)
+            {
+                _spreadTimer += frameTime;
+                if (_spreadTimer > _spreadDelay)
+                {
+                    _spreadTimer -= _spreadDelay;
+
+                    var outerEffects = new HashSet<AreaEffectComponent>(_group.Where(effect => effect.Amount == _amountCounterSpreading));
+                    foreach (var effect in outerEffects)
+                    {
+                        effect.Spread();
+                    }
+
+                    _amountCounterSpreading -= 1;
+                }
+            }
+            // Start counting for _duration after fully spreading
+            else
+            {
+                _lifeTimer += frameTime;
+            }
+
+            // Delete every outer square
+            if (_lifeTimer > _duration)
+            {
+                _spreadTimer += frameTime;
+                if (_spreadTimer > _removeDelay)
+                {
+                    _spreadTimer -= _removeDelay;
+
+                    var outerEffects = new HashSet<AreaEffectComponent>(_group.Where(effect => effect.Amount == _amountCounterRemoving));
+                    foreach (var effect in outerEffects)
+                    {
+                        effect.Kill();
+                    }
+
+                    _amountCounterRemoving += 1;
+                }
+            }
+
+            // Make every square from the group react with the tile and entities
+            _reactionTimer += frameTime;
+            if (_reactionTimer > ReactionDelay)
+            {
+                _reactionTimer -= ReactionDelay;
+                foreach (var effect in _group)
+                {
+                    effect.React(_averageExposures);
```

Comment:
check that effect isn't already deleted, since effect.Kill() in line 202 will trigger the such

Reply:
effect.Kill() already removes the effect from the group so I don't think this is necessary

Reply:
correct